sqlglot/dialects/bigquery.py

```diff
@@ -437,10 +458,265 @@ class Parser(parser.Parser):
+            return self.expression(
+                exp.IfStatement,
+                this=conjunction,
+                if_block=if_block,
+                elseif_blocks=elseif_blocks,
+                else_block=else_block,
+                else_=end_token == TokenType.ELSE,
+            )
+
+        def _parse_statement_block(self, end_tokens: t.Sequence[TokenType]):
+            body = []
+            matched_end_token = None
+            while self._curr:
+                found_end_token = False
+                for end_token in end_tokens:
+                    if self._match(end_token):
+                        matched_end_token = end_token
+                        found_end_token = True
+                        break
+                if found_end_token:
+                    break
+                body.append(self._parse_statement())
+            return body, matched_end_token
+
+        def _parse_begin_block(self):
+            this, end = self._parse_statement_block([TokenType.END])
+            return self.expression(exp.BeginBlock, this=this, end=end is not None)
+
+        def _parse_statement(self) -> t.Optional[exp.Expression]:
+            result = super()._parse_statement()
+            self._match(TokenType.SEMICOLON)
+            return result
+
+        def _parse_call(self):
+            this = self._parse_id_var()
+
+            while self._match(TokenType.DOT):
+                this = self.expression(exp.Dot, this=this, expression=self._parse_id_var())
+
+            call_args = []
+            if self._match(TokenType.L_PAREN):
+                call_args = self._parse_csv(self._parse_expression)
+                self._match(TokenType.R_PAREN)
+
+            return self.expression(exp.Call, this=this, call_args=call_args)
+
+        def _parse_while(self):
+            conjunction = self._parse_conjunction()
+            self._match(TokenType.DO)
+            body, end_token = self._parse_statement_block([TokenType.END_WHILE])
+            return self.expression(
+                exp.While, this=conjunction, body=body, end=end_token is not None
+            )
+
+        def _parse_procedure_body(self, create_token: Token, extend_props: t.Callable):
+            this = self._parse_user_defined_function(kind=create_token.token_type)
+
+            # exp.Properties.Location.POST_SCHEMA ("schema" here is the UDF's type signature)
+            extend_props(self._parse_properties())
+
+            # Big query stored procedures don't support alias,
+            # except for spark ones, which we do not support at the moment.
+            # Reference:
+            # https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#create_procedure
+            begin = self._match(TokenType.BEGIN)
+            procedure_body, end_token = self._parse_statement_block([TokenType.END])
+            return this, procedure_body, begin, end_token is not None
```

Comment:
I'm not sure I understand this change. Weren't we able to parse _valid_ BQ statements before just fine? What's an example that triggered the need to implement this?